### PR TITLE
Education

### PR DESCRIPTION
**Pull request by Vitash1981**

### CODE REVIEW

File: HW3.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+BYE = 'Bye, see you next time! :)'
+WELCOME ='Welcome to a Guessing Word Game!'
+THINK_A_WORD = "Think of a word and enter it's length: "
+POSITIONS_OF_AVAILABLE_LETTERS ='\n If letter(s): "{}" is available in the hidden word, \n please enter comma separated position(s) counting from 0 like 0,2, \n or just press Enter if the letter is absent: '
+EMPTY_STRING = ''
+COMMA = ','
+DOT = '.'
+GUESSED_WORD = 'I want to tell wholw word: "{}"'
+CANT_GUESS = "You are che champions! Can't guess the word matching specified letters"
+NO_TRIES_LEFT = "No more word, so you're probably cheating, stop doing it!"
```
Comment (ValentynTroian) [inline HW3.py:20]: It'd be better to refuse from these variables here. You use them as constants, but the change of their value doesn't change the program logic or calculating. 

Constants are a type of variable that doesn’t change, for example, **PI** constant. They should be used for anything that should not be allowed to change while the program is running.

File: HW3.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+PATH_TO_DATABASE = 'C:\Python\words.txt'
```
Comment (ValentynTroian) [inline HW3.py:21]: The absolute path using isn't good practice. It'd be better to use relative paths. 
Also, I'd change the variable name to PATH_TO_FILE, not to DATABASE.

File: HW3.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+def start_game():
+    if len(sys.argv) < 2:
+        with open(PATH_TO_DATABASE) as f:
+            print(WELCOME)
+            given_word_length = int(input(THINK_A_WORD))
+            search_expr = DOT * given_word_length
+            word_subset = searchUniqueWords(searchWords(f),given_word_length)  # Select words with given length
+            tried_chars = []
+            GameOn = True
+            while GameOn:
```
Comment (ValentynTroian) [inline HW3.py:81]: A little bit more pythonic way:

`while True: `

File: HW3.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        except KeyError:
+            result[char] = 1  # If char not in result dict yet, start counting it
+            sumFreqChars +=1
+    for char in result:
+        freq[char]=round(result[char]*100/sumFreqChars,2)
+
+    result = sorted(freq.items(), key=operator.itemgetter(1), reverse=True)
+    return result
+
+def newGame():
```
Comment (ValentynTroian) [inline HW3.py:64]: It doesn't clear the difference between newGame() and start_game() functions.
It'd better to rename it to "ask_new_game()"

File: HW3.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    answer = input(ONE_MORE_TRY).upper()
+    if answer.startswith(YES):
```
Comment (ValentynTroian) [inline HW3.py:66]: I'd write 

` if answer == 'YES':`

File: HW3.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    sum_freq = 0
+    for key in d:
+        sum_freq += d[key]
+    return sum_freq
+
+def searchUniqueWords(inputList,givenWordLength):
+    listSet = set(inputList)
+    unique_words_list = sorted(list(listSet))
+    possibleWordsMatchedReg = [el.lower() for el in unique_words_list if len(el) == givenWordLength]
+    if possibleWordsMatchedReg==[]:
```
Comment (ValentynTroian) [inline HW3.py:40]: A more Pythonic way:

`if not possibleWordsMatchedReg:`

File: HW3.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    return text
+
+def splitWord(word):
+    return [char for char in word]
+
+def sumFreq(d):
+    sum_freq = 0
+    for key in d:
+        sum_freq += d[key]
+    return sum_freq
```
Comment (ValentynTroian) [inline HW3.py:34]: The function isn't using in the code

File: HW3.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+NO_TRIES_LEFT = "No more word, so you're probably cheating, stop doing it!"
+PATH_TO_DATABASE = 'C:\Python\words.txt'
+
+def searchWords(textFile):
+    text = re.findall (r'\w+', textFile.read().lower())
+    return text
+
+def splitWord(word):
+    return [char for char in word]
+
```
Comment (ValentynTroian) [inline HW3.py:29]: The function isn't using in the code

File: HW3.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            while GameOn:
+                char_counts = getCharFrequency(word_subset)
+                for x in range(len(char_counts)):
+                    char_counts = getCharFrequency(word_subset)
+                    mostFreqChar = char_counts[x][0] # Get most frequent char
+                    if mostFreqChar in tried_chars:
+                        continue  # Try next char if such a char already played
+                    else:
+                        tried_chars.append(mostFreqChar)
+                    occurencesList = input(POSITIONS_OF_AVAILABLE_LETTERS.format(mostFreqChar)).split(COMMA)
```
Comment (ValentynTroian) [inline HW3.py:90]: You can just use:

` .split(", ")`

File: HW3.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+                    if occurencesList != [EMPTY_STRING]:  # if some positions entered, form regexp for further word's filtering
```
Comment (ValentynTroian) [inline HW3.py:91]: A more Pythonic way:

`if occurencesList:`

File: HW3.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+                        if x == len(char_counts):
+                            print(NO_TRIES_LEFT)
+                            sys.exit()
+                        continue  # if nothing entered, than no char the word includes, so repeat with next most popular char
+    else:
+        with open(sys.argv[1], 'r') as f:
+            searchWords(f)                
+
+if __name__ == '__main__':
+    start_game()
```
Comment (ValentynTroian) [inline HW3.py:116]: It'd be better to check if your code meets the PEP8 using pep8 module or [pep8online](http://pep8online.com/checkresult)

Happy coding :)

File: HW4.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
+# readcsv.py -path_csv example.csv -col_name Marker_Strategy
+
+import argparse
+import csv
+import sys
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-path_csv')
+parser.add_argument('-col_name')
```
Comment (ValentynTroian) [inline HW4.py:9]: Your program won't work without **path_csv** and **col_name** arguments, so they are required. So you should use positional arguments. 
If you write "-your_argument", it means that **your_argument** is optional, and your program can work without it

Also it'd be great to add some help information for users.
```
parser.add_argument('path_csv',
                    help='path to input CSV file')
parser.add_argument('col_name',
                    help='CSV column name. If column contains few words, '
                         'use _ between them')
```

File: HW4.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+ns = parser.parse_args(sys.argv[1:])
+
+result = []
+with open(ns.path_csv) as f:
+    reader = csv.DictReader(f)  # read rows into a dictionary format
+    result = [row[ns.col_name.replace('_', ' ')] for row in reader]
```
Comment (ValentynTroian) [inline HW4.py:15]: Your code is replacing col_name for each row.
It seems to me that the following code would be a little bit more efficient:
```
col_name_parsed = ns.col_name.replace('_', ' ')
result = [row[col_name_parsed] for row in reader]
```

File: HW4.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+import argparse
+import csv
+import sys
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-path_csv')
+parser.add_argument('-col_name')
+ns = parser.parse_args(sys.argv[1:])
+
+result = []
```
Comment (ValentynTroian) [inline HW4.py:12]: This assigning is not really needed (it exists the similar one on line 15)

File: HW4.py [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+with open(ns.path_csv) as f:
```
Comment (ValentynTroian) [inline HW4.py:13]: The program will crash if the file doesn't exist.
Probably it'd be better to use "try-except" construction.

Comment (Vitash1981) [reply HW4.py:13]: Thank you for your advise :)

**Pull request merged**